### ISSUE TITLE
Allow 'manual' population

### ISSUE THREAD

**Issue by fiznool** [Automattic/mongoose#2370]
This is the same concept as #1344 but with an easier example to digest.

Here is an example schema to illustrate the feature request.

``` js
var EmbeddedSchema = new mongoose.Schema({});
var Embedded = mongoose.model('Embedded', EmbeddedSchema);

var MainSchema = new mongoose.Schema({
  embedded: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Embedded'
  }
});
var Main = mongoose.model('SetValueSpecMain', MainSchema);

var embedded = new Embedded();
var main = new Main();

// Setting via 'public' API just sets the ID
main.set('embedded', embedded);
expect(main.embedded).toEqual(embedded._id);

// Setting via 'private' API allows assignment of real object
main.setValue('embedded', embedded);
expect(main.embedded).toBe(embedded);
```

Essentially, with a document `embedded`, I'd like to be able to set this doc to the `main.embedded` property. Sort of 'manual' population, or population with an existing document. Using `setValue` achieves the goal, but this function is marked as private, and I don't have a deep enough knowledge of mongoose to understand why this is.

This feature becomes useful (in real life) when the `embedded` document was already fetched in a previous query, so this removes the need to go back to the database through a call to `populate()`.

**Comment by shadow0162** [comment]
it's there a way to popluated by manually?

**Comment by vkarpov15** [comment]
Yep you can actually set `doc.populatedPath = childDoc` in 4.0.0

**Comment by sabrehagen** [comment]
Re "Yep you can actually set doc.populatedPath = childDoc in 4.0.0", does this work for schemas of the below form, i.e. those without an explicit ref field?

```
var socialLinkSchema = Schema({
    parent : { type : ObjectId }
});
```
I'm trying to assign a document to the parent field of my schema and having only the id plucked out in 4.0.7.

**Comment by vkarpov15** [comment]
@sabrehagen nope. This feature is hidden behind [a pretty rigid check that makes sure that the `ref` option is equal to the modelName of `childDoc`](https://github.com/Automattic/mongoose/blob/d6077c9cde84b05af506f49db33c79e2ea55c56c/lib/document.js#L596-601)

**Comment by adambuczynski** [comment]
@vkarpov15 I am having trouble getting this to work for nested schema's, and had to resolve to an ugly hack to make it work, manually setting `doc.constructor.modelName` to the same value I set for `ref`, in order to bypass the rigid check you pointed at.

The use case is that I'm using a nested schema for child documents within a parent document, let's call it A, and then trying to manually populate references in another document B to the child documents of A.

It seems impossible to do this via the .populate() method, see https://github.com/Automattic/mongoose/issues/3312, so I figured I'd try manual population. But then this also doesn't seem possible unless I use the aforementioned hack. Any thoughts?
